### PR TITLE
Improve Contributing doc

### PR DESCRIPTION
**Pull request by jtoar**
> Link to [Contributing](https://deploy-preview-174--redwoodjs.netlify.app/docs/contributing) doc
> Link to tracking issue #156 

This PR shows a WIP edit to the contributing doc. Wanted to get some early feedback to see if I was heading in the right direction.

Some of the text is just copied over from the doc on the repo; I'll make it more distinct where it should be but just using to hold place for the time being.

@thedavidprice One question that came up when editing this was: should the doc on contributing docs be in this doc? (Sorry for using "doc" three times.) Right now I'm putting it at the end (## Contributing Docs). I think it makes sense here but what do you think?

### CODE REVIEW

File: docs/contributing.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+The Tutorial is a standalone document that serves a specific purpose as an introduction to Redwood, an aspirational roadmap, and an example of developer experience. As such, it's distinct from the categories mentioned, although it's most similar to Cookbooks.
+
+#### Docs
+
+Docs are explanation-driven "how-to" content. They're more direct and to-the-point than the tutorial and cookbooks. The idea is much more about finding something or "getting something done" than any kind of "learning journey". 
+<!-- like the tutorial, they're all about Redwood, whereas cookbooks can be about integrating third-parties with a little bit of slapstick humor along the way.  -->
+
+Before you take on a doc, you should read [Form](https://redwoodjs.com/docs/form) and [Redwood Router](https://redwoodjs.com/docs/redwood-router); they have the kind of content you should be striving for. They're comprehensive yet conversational. 
+
+In general, don't be afraid to go into too much detail. We'd rather you err on the side of too much than too little. One tip for finding good content is searching the forum/github for prior art--what are people talking about where this comes up?
```
Comment (Tobbe) [inline docs/contributing.md:86]:
```suggestion
In general, don't be afraid to go into too much detail. We'd rather you err on the side of too much than too little. One tip for finding good content is searching the forum/github for prior art&mdash;what are people talking about where this comes up?
```

File: docs/contributing.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+<!-- damn adverbs -->
+Cookbooks are  tutorial-style content focused on a specific problem-solution. They usually have a beginner in mind (if not, they should indicate that they don't&mdash;put 'Advanced' or 'Deep-Dive', etc., in the title or introduction). Cookbooks may include some explanatory text as asides, but they shouldn't be the majority of the content.
+
+#### Making a Doc Searchable
+
+[TODO] Searchability in and outside redwoodjs.com; outside meaning google, etc., inside meaning algolia.
+
+### "Contributing" Docs (redwoodjs/redwood)
+
+These docs are in the Framework repo, redwoodjs/redwood, and explain how to contribute to Redwood packages. They're the docs linked to the in table above.
```
Comment (Tobbe) [inline docs/contributing.md:101]:
```suggestion
These docs are in the Framework repo, redwoodjs/redwood, and explain how to contribute to Redwood packages. They're the docs linked to in the table above. 
```

**Comment by thedavidprice** [comment]
@Tobbe thanks for the help here!

@jtoar this is looking fantastic 🚀

### Contributing Code

This content is good content, so keep it here! But also need to mention and link to the main Framework CONTRIBUTING doc as it's needed to set up the local dev, etc.

### "Developing" Docs (redwoodjs.com)

Maybe add a sentence here that each of these types of docs are in this repo within respective directory, docs/ cookbooks/ etc.

### Making a Doc Searchable
>[TODO] Searchability in and outside redwoodjs.com; outside meaning google, etc., inside meaning algolia.

Let's call this **"Making a Doc Findable"**
In general, step one is to "Consider the word(s) or phrase(s) people will use to _find_ the information. If you didn't know much about Redwood Configuration, for example, a menu title `redwood.toml` wouldn't make any sense even though that **is** the main configuration file. You'd probably look for "Redwood Config" or "Settings" in the main menu. Or type "how to change Redwood App settings" in the site Search or Google. So make a list of 2-3 keywords or phrases, and try to use each of those in step 2.

Step two is to use those keywords/phrases in 3 sections, in this order of priority:
1. the right column menu title (for a page)
2. the page title or the section title 
3. the introductory paragraph of the page _or_ the description of the page section

### "Contributing" Docs (redwoodjs/redwood)
This is good an brief, so I think it's ok to leave here. More importantly, this should likely be included directly in the Framework CONTRIBUTING with a link from there referencing the "Contributing Docs" section of this for approach/style information. 

I know it feels redundant. No right answer at this time. Being more thorough is better than people not finding info at this time. And in this case there's no need to have the content kept "in since" if you do some copy+paste duplication.

### @redwoodjs/auth and @redwoodjs/router
I would remove this section from here and put it directly in both of the redwoodjs.com _placeholder_ docs for `redwoodjs.com/docs/authentication.md` and `https://github.com/redwoodjs/redwoodjs.com/blob/master/docs/router.md`

---

This is looking great. Let's get it 🚢'ed!

**Comment by thedavidprice** [comment]
Another note on the Framework docs re: package READMEs --> docs should only live in the package root and within the README. Let's standardize this as I'm seeing the occasional README pop up in a sub-directory.

**Comment by jtoar** [comment]
@thedavidprice Changes made, but could you clarify this part:

> 1. the right column menu title (for a page)
> 2. the page title or the section title
> 3. the introductory paragraph of the page or the description of the page section

If I have this right: 1. is the heading under the "On this page" nav, 2. is the page's title at the top (section title is?), and 3. no confusion here. But was wondering why you excluded the menu title? As in, the heading in the left nav.

**Comment by thedavidprice** [comment]
@jtoar Ah, sorry, I biffed that 🤦‍♂️

>the ~~right~~ **left** column menu title (for a page)

^^ yes, definitely the menu title is priority 1

> the page title or the section title

^^ Page title may/may not be same as main menu title but if needed can be different (per discussion as before); but section is the same as right nav

>the introductory paragraph of the page or the description of the page section

^^ roger

**Comment by jtoar** [comment]
@thedavidprice This one's ready for review again: https://deploy-preview-174--redwoodjs.netlify.app/docs/contributing. A few last things I caught:
- Had to change some of the titles in "On this page" because they weren't working for whatever reason, with quotes and parenthesis and all.
- Changed Cookbooks to Cookbook Recipes since there's actually one Cookbook with a bunch of recipes.

File: docs/contributing.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
+# Contributing
+
+Love Redwood and want to get involved? You’re in the right place! There's several ways you can contribute to Redwood:
```
Comment (Tobbe) [inline docs/contributing.md:3]:
```suggestion
Love Redwood and want to get involved? You’re in the right place! There are several ways you can contribute to Redwood:
```

File: docs/router.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
...
+Generally, there's a doc context rule--developing docs go on the website, contributing docs go on the repo. But since `@redwoodjs/router` can be used outside Redwood apps, we have to break this rule since it's not just about a Redwood user's experience anymore.
 
-So, if you want to edit the Redwood Router doc, go to the repo: https://github.com/redwoodjs/redwood/blob/master/packages/router/README.md.
\ No newline at end of file
+> Why design our packages so that they can work outside of Redwood apps too? Well, one of the reasons we made Redwood was to inject more integration into the JS ecosystem. And if we can accelerate this phenemenon, why wouldn't we?
+
+So unless we want to face the perils of synchronizing changes across two docs in two places, the package's README kind of needs to have it all.
+
+This is all to say that if you  want to edit these docs (and thank you if you do!), you'll want to go to the [redwoodjs/redwood](https://github.com/redwoodjs/redwood) repo and submit your PR there.
```
Comment (Tobbe) [inline docs/router.md:9]:
```suggestion
This is all to say that if you want to edit these docs (and thank you if you do!), you'll want to go to the [redwoodjs/redwood](https://github.com/redwoodjs/redwood) repo and submit your PR there.
```

File: docs/authentication.md [hunk 5]
```diff
@@ -1,5 +1,9 @@
...
+Generally, there's a doc context rule--developing docs go on the website, contributing docs go on the repo. But since `@redwoodjs/auth` can be used outside Redwood apps, we have to break this rule since it's not just about a Redwood user's experience anymore.
 
-So, if you want to edit the redwood auth doc, go to the repo: https://github.com/redwoodjs/redwood/blob/master/packages/auth/README.md.
\ No newline at end of file
+> Why design our packages so that they can work outside of Redwood apps too? Well, one of the reasons we made Redwood was to inject more integration into the JS ecosystem. And if we can accelerate this phenemenon, why wouldn't we?
+
+So unless we want to face the perils of synchronizing changes across two docs in two places, the package's README kind of needs to have it all.
+
+This is all to say that if you want to edit these docs (and thank you if you do!), you'll want to go to the [redwoodjs/redwood](https://github.com/redwoodjs/redwood) repo and submit your PR there.
```
Comment (thedavidprice) [inline docs/authentication.md:9]:
```suggestion
This is all to say that if you want to edit this doc (and thank you if you do!), use the following link to edit the Auth package README in the `redwoodjs/redwood` repo and submit a PR there:
https://github.com/redwoodjs/redwood/blob/master/packages/auth/README.md
```

File: docs/router.md (hunk 4 above)
Comment (thedavidprice) [inline docs/router.md:9]:
```suggestion
This is all to say that if you want to edit this doc (and thank you if you do!), use the following link to edit the Router package README in the `redwoodjs/redwood` repo and submit a PR there:
https://github.com/redwoodjs/redwood/blob/master/packages/router/README.md
```

File: docs/contributing.md (hunk 3 above)
Comment (thedavidprice) [inline docs/contributing.md:3]:
```suggestion
Love Redwood and want to get involved? You’re in the right place! There are several ways you can contribute to Redwood:
```

File: docs/contributing.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+What you want to do not on the roadmap? Well, still go for it! We love spikes and proof-of-concepts. And if you have a question, just ask!
+
+|Package|Description|
+|:-|:-|
+|[`@redwoodjs/api`](https://github.com/redwoodjs/redwood/blob/master/packages/api/README.md)|Exposes functions to build the GraphQL API and provides services with `context`|
+|[`@redwoodjs/auth`](https://github.com/redwoodjs/redwood/blob/master/packages/auth/README.md#contributing)|A lightweight wrapper around popular SPA authentication libraries|
+|[`@redwoodjs/cli`](https://github.com/redwoodjs/redwood/blob/master/packages/cli/README.md)|All the commands for Redwood's built-in CLI|
+|[`@redwoodjs/core`](https://github.com/redwoodjs/redwood/blob/master/packages/core/README.md)|Defines babel plugins and config files|
+|[`@redwoodjs/create-redwood-app`](https://github.com/redwoodjs/redwood/blob/master/packages/create-redwood-app/README.md)|Enables `yarn create redwood-app`&mdash;downloads the latest release of Redwood and extracts it into the supplied directory|
+|[`@redwoodjs/dev-server`](https://github.com/redwoodjs/redwood/blob/master/packages/dev-server/README.md)|Maps lambda functions to URLs|
```
Comment (thedavidprice) [inline docs/contributing.md:29]:
```suggestion
|[`@redwoodjs/dev-server`](https://github.com/redwoodjs/redwood/blob/master/packages/dev-server/README.md)|Configuration for the local development server|
```

File: docs/contributing.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+Before you take on a doc, you should read [Form](https://redwoodjs.com/docs/form) and [Redwood Router](https://redwoodjs.com/docs/redwood-router); they have the kind of content you should be striving for. They're comprehensive yet conversational. 
+
+In general, don't be afraid to go into too much detail. We'd rather you err on the side of too much than too little. One tip for finding good content is searching the forum and repo for "prior art"&mdash;what are people talking about where this comes up?
+
+#### Cookbook Recipes
+
+Cookbook Recipes are tutorial-style content focused on a specific problem-solution. They usually have a beginner in mind (if not, they should indicate that they don't&mdash;put 'Advanced' or 'Deep-Dive', etc., in the title or introduction). Cookbook Recipes may include some explanatory text as asides, but they shouldn't be the majority of the content.
+
+#### Making a Doc Findable
```
Comment (thedavidprice) [inline docs/contributing.md:70]: well done on this section 🚀

File: docs/contributing.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+## Contributing Code
+
+Redwood's composed of many packages that are designed to work together. Some of these packages are designed to be used outside Redwood too!
+
+Before you start contributing, you'll want to set up your local development environment. The Redwood repo's top-level [contributing guide](https://github.com/redwoodjs/redwood/blob/master/CONTRIBUTING.md#local-development) walks you through this. Make sure to give it an initial read.
+
+For details on contributing to a specific package, see the package's README (links provided in the table below). Each README has a section named Roadmap. If you want to get involved but don't quite know how, the Roadmap's a good place to start&mdash;see anything that interests you? Go for it! And be sure to let us know&mdash;you don't have to have a finished product before opening an issue or pull request. In fact, we're big fans of [Readme Driven Development](https://tom.preston-werner.com/2010/08/23/readme-driven-development.html).
```
Comment (Tobbe) [inline docs/contributing.md:18]: I thought the use of em-dash looked weird. But I've never used it myself, so didn't comment on it the first time I saw it. But now, looking through this again I once again got caught on the em-dash. So I had to look it up. Found this interesting article about it https://www.punctuationmatters.com/hyphen-dash-n-dash-and-m-dash/  Looks like you're technically correct in your usage. But his section on "Breaking the rules" was pretty much spot on how I felt about it :) 
> I usually use ‘space-n-dash-space’ instead of the m-dash – just to keep everyone happy. 

Some more reading about it tells me that we don't use m-dash in Swedish. That's probably why I think it looks weird. The Swedish wikipedia page even notes that (paraphrasing) "The em-dash isn't commonly used in modern Swedish, but it is used in English". Kind of like how English (sometimes) uses double space after period, but we never do.

Basically this is just me thinking out loud. Do what you want with it. 🙂

Comment (jtoar) [reply docs/contributing.md]: Double spaces after periods :scream: I remember doing that in middle school. Never again. And thanks for the back-and-forth. That's super interesting/something I would've never known. 

Em-dashes are (another) one of my weaknesses. I removed the first one&mdash;it doesn't have to be there, especially since it's so closely followed by another. 

I tend to use the em-dash for the second function listed in the article:

> Separates an inserted thought or clause from the main clause

Never heard of the space-en-dash-space, and can't say I think it's better than just the em yet. But I'll try not to be as liberal in my use of them going forward. 

Literary aside: if you really want your punctuation mind to be blown, read Cormac McCarthy's Blood Meridian. He scarcely uses anything but a period. It's amazing. And also (as the title would suggest) terribly bloody/gruesome.

Comment (thedavidprice) [reply docs/contributing.md]: I ❤️ em dash.

Also, I ❤️ Cormac McCarthy.

Comment (Tobbe) [reply docs/contributing.md]: Just searched online, and they have Blood Meridian at my local library. But, of course it's in Swedish. I'm sure it's still a great story, but I wonder if the translator managed to keep the punctuation mind-blowery...
